include/zephyr/math/q_ops.h

```diff
@@ -0,0 +1,139 @@
+		 q1_0_31_t : 1)
+
+#define q_assign(qval, x)                                                                          \
+	do {                                                                                       \
+		int64_t intermediate =  z_q_val_to_intermediate(x);                                \
+		intermediate = intermediate >> z_q_right_shift_bits(qval);                         \
+        	__ASSERT_NO_MSG(intermediate <= INT32_MAX && intermediate >= INT32_MIN);           \
+		qval.value = intermediate & 0xffffffff;                                            \
+	} while (false)
+
+#define q_add(qval, x)                                                                             \
+	do {                                                                                       \
+                int64_t lhs = (int64_t)qval.value;                                                 \
+                int64_t rhs = z_q_val_to_intermediate(x) >> z_q_right_shift_bits(qval);            \
+		int64_t intermediate = lhs + rhs;                                                  \
+        	__ASSERT_NO_MSG(intermediate <= INT32_MAX && intermediate >= INT32_MIN);           \
+		qval.value = intermediate & 0xffffffff;                                            \
+	} while (false)
+
+#define q_sub(qval, x)                                                                             \
+	do {                                                                                       \
+                int64_t lhs = (int64_t)qval.value;                                                 \
+                int64_t rhs = z_q_val_to_intermediate(x) >> z_q_right_shift_bits(qval);            \
+		int64_t intermediate = lhs - rhs;                                                  \
+        	__ASSERT_NO_MSG(intermediate <= INT32_MAX && intermediate >= INT32_MIN);           \
+		qval.value = intermediate & 0xffffffff;                                            \
+	} while (false)
+
+#define q_mul(qval, x)                                                                             \
+	do {                                                                                       \
+                int64_t lhs = (int64_t)qval.value;                                                 \
+		int64_t rhs = z_q_val_to_intermediate(x) >> z_q_right_shift_bits(qval);            \
+        	int64_t intermediate =  (lhs * rhs) >> _Generic((qval),                            \
+							q1_23_8_t : 8,                             \
+							q1_15_16_t : 16,                           \
+							q1_0_31_t : 31);                           \
+        	__ASSERT_NO_MSG(intermediate <= INT32_MAX && intermediate >= INT32_MIN);           \
+		qval.value = intermediate & 0xffffffff;                                            \
+	} while (false)
+
+#define q_div(qval, x)                                                                             \
```

Comment:
You can use gcc macro expressions to provide returns instead of mutating the given variable, there's really no other way to do that otherwise that I'm aware of (I don't know everything). If people are using non-gcc compatible (gcc/clang) then perhaps this subsys won't be built or allowed, and that's ok I think.

Reply:
Just saw this, I'll make the change to update to return a value instead of mutating tomorrow.

Reply:
I'm going to hold off doing more work on this until the discussion over CMSIS-DSP is resolved.